File: 2020/16.py
import sys

def ParseValidRanges(part):
    ranges = {}
    for line in part.split('\n'):
        key, values = line.split(': ')
        ranges[key] = []
        for part in values.split(' or '):
            lo, hi = map(int, part.split('-'))
            assert lo <= hi
            ranges[key].append((lo, hi))
    return ranges

def ParseYourTicket(part):
    first, second = part.split('\n')
    assert first == 'your ticket:'
    return list(map(int, second.split(',')))

def ParseNearbyTickets(part):
    first, *rest = part.split('\n')
    assert first == 'nearby tickets:'
    return [list(map(int, line.split(','))) for line in rest]

# Parse input
part1, part2, part3 = sys.stdin.read().strip().split('\n\n')
ranges = ParseValidRanges(part1)            # key -> [(lo, hi)]
my_ticket = ParseYourTicket(part2)          # [num]
nearby_tickets = ParseNearbyTickets(part3)  # [[num]]

# Part 1: filter out invalid tickets
valid_tickets = []
invalid_sum = 0
for nums in nearby_tickets:
    valid = True
    for num in nums:
        if not any(lo <= num <= hi for bounds in ranges.values() for lo, hi in bounds):
            valid = False
            invalid_sum += num
    if valid:
        valid_tickets.append(nums)
print(invalid_sum)


def CalculatePositions():
    # For each position, keep track of set of possible keys.
    possibilities = [set(ranges) for _ in my_ticket]  # [set([key])]

    # Pass 1: remove all keys that would make a ticket invalid.
    for nums in valid_tickets:
        for i, num in enumerate(nums):
            for key in list(possibilities[i]):
                if not any(lo <= num <= hi for lo, hi in ranges[key]):
                    possibilities[i].remove(key)

    # Pass 2: find uniquely-determined keys and remove it as a possibility
    # from oher positions (like naked singles in Sudoku). Repeat recursively.
    positions = {}  # key (str) -> position (int)
    def Check(i):
        assert len(possibilities[i]) > 0
        if len(possibilities[i]) > 1:
            return
        key, = possibilities[i]
        if key in positions:
            assert positions[key] == i
            return
        positions[key] = i
        for j, keys in enumerate(possibilities):
            if i != j and key in keys:
                keys.remove(key)
                Check(j)
    for i in range(len(possibilities)):
        Check(i)
    return positions

# Part 2: find which keys map to which field indices, then compute
# the product of the fields with key prefix "departure".
positions = CalculatePositions()
departure_product = 1
for key in ranges:
    if key.startswith('departure '):
        departure_product *= my_ticket[positions[key]]
print(departure_product)

File: 2020/haskell/16.hs
import Data.List
import Data.List.Split
import PairWiseAssociation

data Range = Range Int Int deriving Show

data Field = Field {
    fieldKey :: String,
    fieldRanges :: [Range]} deriving Show

instance Eq Field where
    (==) Field{fieldKey=k1} Field{fieldKey=k2} = k1 == k2

type Ticket = [Int]

parseRange :: String -> Range
parseRange word | a <= b = Range a b where [a, b] = map read $ splitOn "-" word

parseField :: String -> Field
parseField line = Field key ranges
    where
        [key, rangeDefs] = splitOn ": " line
        ranges = map parseRange $ splitOn " or " rangeDefs

parseTicket :: String -> Ticket
parseTicket line = map read $ splitOn "," line

inRange :: Int -> Range -> Bool
inRange x (Range a b) = a <= x && x <= b

inAnyRange :: Int -> [Range] -> Bool
inAnyRange value = any (inRange value)

invalidValues :: [Field] -> Ticket -> [Int]
invalidValues fields = filter (not . (`inAnyRange` ranges))
    where
        ranges = concatMap fieldRanges fields

totalErrors :: [Field] -> [Ticket] -> Int
totalErrors fields tickets = sum $ concatMap (invalidValues fields) tickets

isValidTicket :: [Field] -> Ticket -> Bool
isValidTicket fields = null . invalidValues fields

-- Uses the ranges from `fields` and the values from `tickets` to determine
-- which field corresponds with which position on the ticket, and returns a
-- list of fields paired with corresponding elements from `xs`.
--
-- For example, associateFields fields tickets [0..] will return a list of
-- (field, position) pairs, but not necessarily ordered by position.
--
-- In the solution below, we'll directly associate fields with ticket values,
-- bypassing the need to index lists by positions.
--
-- (We assume all field positions can be uniquely determined, which is true for
-- the given test data, but not strictly necessary for the problem, since we
-- only need to identify a subset of fields.)
associateFields :: [Field] -> [Ticket] -> [a] -> [(Field, a)]
associateFields fields tickets xs
    = solvePairs $ zip (map possibilities $ transpose tickets) xs
    where
        -- Returns a list of possible fields based on initial ticket values.
        possibilities :: [Int] -> [Field]
        possibilities values = filter possible fields
            where possible field = all (`inAnyRange` (fieldRanges field)) values

main = do
    input <- getContents
    let [part1, part2, part3] = splitOn [[]] (lines input)
    let fields = map parseField part1
    let yourTicket = let ["your ticket:", def] = part2 in parseTicket def
    let nearbyTickets = let ("nearby tickets:":defs) = part3 in map parseTicket defs

    -- Part 1: count sum of invalid ticket values.
    print $ totalErrors fields nearbyTickets

    -- Part 2: identifiy field positions and calculate product of "departure" fields.
    let validTickets = filter (isValidTicket fields) nearbyTickets
    let departureValues = [value |
            (field, value) <- associateFields fields validTickets yourTicket,
            "departure " `isPrefixOf` (fieldKey field)]
    print $ product departureValues
